fe/src/main/java/org/apache/doris/analysis/StmtRewriter.java

```diff
@@ -953,5 +957,81 @@ private static Expr createJoinConjunct(Expr exprWithSubquery, InlineViewRef inli
         smap.put(subquery, subquerySubstitute);
         return exprWithSubquery.substitute(smap, analyzer, false);
     }
+
+    /** rewrite subquery in case when to an inline view
+     *  subquery in case when statement like
+     *
+     * SELECT CASE
+     *         WHEN (
+     *             SELECT COUNT(*) / 2
+     *             FROM t
+     *         ) > k4 THEN (
+     *             SELECT AVG(k4)
+     *             FROM t
+     *         )
+     *         ELSE (
+     *             SELECT SUM(k4)
+     *             FROM t
+     *         )
+     *     END AS kk4
+     * FROM t;
+     * this statement will be rewrite to
+     *
+     * SELECT CASE
+     *         WHEN t1.a > k4 THEN t2.a
+     *         ELSE t3.a
+     *     END AS kk4
+     * FROM t, (
+     *         SELECT COUNT(*) / 2 AS a
+     *         FROM t
+     *     ) t1,  (
+     *         SELECT AVG(k4) AS a
+     *         FROM t
+     *     ) t2,  (
+     *         SELECT SUM(k4) AS a
+     *         FROM t
+     *     ) t3;
+     */
+    private static void rewriteCaseWhenSubqueries(SelectStmt stmt, Analyzer analyzer) throws AnalysisException {
+        for (SelectListItem item: stmt.getSelectList().getItems()) {
+            if (!(item.getExpr() instanceof CaseExpr)) {
+                continue;
+            }
+            if (!item.getExpr().contains(Predicates.instanceOf(Subquery.class))) {
+                continue;
+            }
+            item.setExpr(rewriteCaseWhenSubquery(item.getExpr(), stmt, analyzer));
+        }
+
+    }
+
+    private static Expr rewriteCaseWhenSubquery(Expr expr, SelectStmt stmt, Analyzer analyzer)
+            throws AnalysisException {
+        if (expr instanceof Subquery) {
+            if (!(((Subquery) expr).getStatement() instanceof SelectStmt)) {
+                throw new AnalysisException("only support select subquery in case statement.");
+            }
+            SelectStmt subquery = (SelectStmt) ((Subquery) expr).getStatement();
+            if (subquery.getSelectList().getItems().size() != 1) {
+                throw new AnalysisException("only support scala select subquery in case statement.");
+            }
+            subquery.reset();
+            String alias = stmt.getTableAliasGenerator().getNextAlias();
+            String colAlias = stmt.getColumnAliasGenerator().getNextAlias();
+            InlineViewRef inlineViewRef = new InlineViewRef(alias, subquery, Arrays.asList(colAlias));
+            try {
+                inlineViewRef.analyze(analyzer);
+            } catch (UserException e) {
+                throw new AnalysisException(e.getMessage());
+            }
+            stmt.fromClause_.add(inlineViewRef);
+            expr = new SlotRef(new TableName(null, alias), colAlias);
```

Comment:
The table name could be fetch from inline view ref

Reply:
first parameter is database name